flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/datasource/transactions/xa/XaTransaction.java

```diff
@@ -0,0 +1,295 @@
+                result.failedTransiently(x.withAttemptsIncremented(), e);
+            } catch (Exception e) {
+                result.failed(x, e);
+            }
+        }
+        result.getForRetry().addAll(xids);
+        result.throwIfAnyFailed("commit");
+        throwIfAnyReachedMaxAttempts(result, maxCommitAttempts);
+        result.getTransientFailure()
+                .ifPresent(
+                        f ->
+                                LOG.warn(
+                                        "failed to commit {} transactions out of {} (keep them to retry later)",
+                                        result.getForRetry().size(),
+                                        origSize,
+                                        f));
+        return result;
+    }
+
+    private XaTransactionResult<TransactionId> failOrRollback(Collection<TransactionId> xids) {
+        XaTransactionResult<TransactionId> result = new XaTransactionResult<>();
+        if (xids.isEmpty()) {
+            return result;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("rolling back {} transactions: {}", xids.size(), xids);
+        }
+        for (TransactionId x : xids) {
+            try {
+                xaConnectionProvider.failAndRollback(x);
+                result.succeeded(x);
+            } catch (TransientXaException e) {
+                LOG.info("unable to fail/rollback transaction, xid={}: {}", x, e.getMessage());
+                result.failedTransiently(x, e);
+            } catch (Exception e) {
+                LOG.warn("unable to fail/rollback transaction, xid={}: {}", x, e.getMessage());
+                result.failed(x, e);
+            }
+        }
+        if (!result.getForRetry().isEmpty()) {
+            LOG.info("failed to roll back {} transactions", result.getForRetry().size());
+        }
+        return result;
+    }
+
+    private void recoverAndRollback() {
+        Collection<Xid> recovered = xaConnectionProvider.recover();
+        if (recovered.isEmpty()) {
+            return;
+        }
+        LOG.warn("rollback {} recovered transactions", recovered.size());
+        for (Xid xid : recovered) {
+            if (baseTransaction.belongsTo(xid)) {
+                try {
+                    xaConnectionProvider.rollback(xid);
+                } catch (Exception e) {
+                    LOG.info("unable to rollback recovered transaction, xid={}", xid, e);
+                }
+            }
+        }
+    }
+
+    private void throwIfAnyReachedMaxAttempts(
+            XaTransactionResult<TransactionId> result, int maxAttempts) {
+        List<TransactionId> reached = null;
+        for (TransactionId x : result.getForRetry()) {
+            if (x.getAttempts() >= maxAttempts) {
+                if (reached == null) {
+                    reached = new ArrayList<>();
+                }
+                reached.add(x);
+            }
+        }
+        if (reached != null) {
```

Comment:
Maybe `Precondition` utils is the better option.

Reply:
I would need to change to checkArgument(reached == null, "error message") and this will throw a IllegalArgumentException.

Its not really a bad argument, I would say that I prefer the way it is